crates/vm/src/eval.rs

```diff
@@ -1,382 +1,60 @@
-
-                    ExprOpt::Query(Box::new(source))
-                }
-                CrudExpr::Insert { source, rows: data } => {
-                    let source = build_source(source);
-                    let mut rows = Vec::with_capacity(data.len());
-                    for x in data {
-                        let mut row = Vec::with_capacity(x.len());
-                        for v in x {
-                            match v {
-                                FieldExpr::Name(x) => {
-                                    todo!("Deal with idents in insert?: {}", x)
-                                }
-                                FieldExpr::Value(x) => {
-                                    row.push(x);
-                                }
-                            }
-                        }
-                        rows.push(ProductValue::new(&row))
-                    }
-                    ExprOpt::Crud(Box::new(CrudExprOpt::Insert { source, rows }))
-                }
-                CrudExpr::Update { delete, assignments } => {
-                    let delete = build_query_opt(delete);
 
-                    ExprOpt::Crud(Box::new(CrudExprOpt::Update { delete, assignments }))
-                }
-                CrudExpr::Delete { query } => {
-                    let query = build_query_opt(query);
-
-                    ExprOpt::Crud(Box::new(CrudExprOpt::Delete { query }))
-                }
-                CrudExpr::CreateTable { table } => ExprOpt::Crud(Box::new(CrudExprOpt::CreateTable { table })),
-                CrudExpr::Drop {
-                    name,
-                    kind,
-                    table_access,
-                } => ExprOpt::Crud(Box::new(CrudExprOpt::Drop {
-                    name,
-                    kind,
-                    table_access,
-                })),
-            }
-        }
-        x => {
-            todo!("{:?}", x)
-        }
-    }
-}
-
-/// First pass:
-///
-/// Compile the [Expr] into a type-annotated AST [Tree<ExprOpt>].
-///
-/// Then validate & type-check it.
-#[tracing::instrument(skip_all)]
-pub fn optimize<P: ProgramVm>(p: &mut P, code: Expr) -> Result<ExprOpt, ErrorType> {
-    let result = build_typed(p, code);
-    check_types(&mut p.env_mut().ty, &result)?;
-
-    Ok(result)
-}
-
-fn collect_vec<P: ProgramVm>(p: &mut P, iter: impl ExactSizeIterator<Item = ExprOpt>) -> Result<Vec<Code>, ErrorVm> {
-    let mut code = Vec::with_capacity(iter.len());
-
-    for x in iter {
-        code.push(compile(p, x)?);
-    }
-    Ok(code)
-}
-
-fn collect_map<P: ProgramVm>(
-    p: &mut P,
-    iter: impl ExactSizeIterator<Item = (String, ExprOpt)>,
-) -> Result<HashMap<String, Code>, ErrorVm> {
-    let mut code = HashMap::with_capacity(iter.len());
-
-    for (name, x) in iter {
-        code.insert(name, compile(p, x)?);
-    }
-    Ok(code)
-}
-
-fn compile_query(q: QueryExprOpt) -> QueryCode {
+pub(crate) fn compile_query(q: QueryExpr) -> QueryCode {
```

Comment:
So we're not constructing a `QueryExprOpt` anymore?

Reply:
Oh, correct, I miss that and the dependant code